File: csharp/src/impl/ScalableSet.cs
using System;
using System.Collections;
using System.Collections.Generic;

namespace Volante.Impl
{
    class ScalableSet<T> : PersistentCollection<T>, ISet<T> where T : class,IPersistent
    {
        ILink<T> link;
        ISet<T> pset;
        const int BTREE_THRESHOLD = 128;

        internal ScalableSet(DatabaseImpl db, int initialSize)
            : base(db)
        {
            if (initialSize <= BTREE_THRESHOLD)
                link = db.CreateLink<T>(initialSize);
            else
                pset = db.CreateBtreeSet<T>();
        }

        ScalableSet() { }

        public override int Count
        {
            get
            {
                if (pset != null)
                    return pset.Count;
                else
                    return link.Count;
            }
        }

        public override void Clear()
        {
            if (pset != null)
                pset.Clear();
            else
            {
                link.Clear();
                Modify();
            }
        }

        public override bool Contains(T o)
        {
            if (pset != null)
                return pset.Contains(o);
            else
                return link.Contains(o);
        }

        public T[] ToArray()
        {
            if (pset != null)
                return pset.ToArray();
            else
                return link.ToArray();
        }

        public override IEnumerator<T> GetEnumerator()
        {
            if (pset != null)
                return pset.GetEnumerator();
            else
                return link.GetEnumerator();
        }

        IEnumerator IEnumerable.GetEnumerator()
        {
            return GetEnumerator();
        }

        public override void Add(T o)
        {
            if (pset != null)
            {
                pset.Add(o);
                return;
            }

            if (link.IndexOf(o) >= 0)
                return;

            if (link.Count <= BTREE_THRESHOLD)
            {
                Modify();
                link.Add(o);
                return;
            }

            pset = ((DatabaseImpl)Database).CreateBtreeSet<T>();
            for (int i = 0, n = link.Count; i < n; i++)
            {
                pset.Add(link[i]);
            }
            link = null;
            Modify();
            pset.Add(o);
        }

        public override bool Remove(T o)
        {
            if (pset != null)
                return pset.Remove(o);

            int i = link.IndexOf(o);
            if (i < 0)
                return false;

            link.RemoveAt(i);
            Modify();
            return true;
        }

        public bool ContainsAll(ICollection<T> c)
        {
            if (pset != null)
                return pset.ContainsAll(c);

            foreach (T o in c)
            {
                if (!Contains(o))
                    return false;
            }
            return true;
        }

        public bool AddAll(ICollection<T> c)
        {
            if (pset != null)
                return pset.AddAll(c);

            bool modified = false;
            foreach (T o in c)
            {
                if (!Contains(o))
                {
                    modified = true;
                    Add(o);
                }
            }
            return modified;
        }

        public bool RemoveAll(ICollection<T> c)
        {
            if (pset != null)
                return pset.RemoveAll(c);

            bool modified = false;
            foreach (T o in c)
            {
                modified |= Remove(o);
            }
            return modified;
        }

        public override bool Equals(object o)
        {
            if (o == this)
                return true;

            ISet<T> s = o as ISet<T>;
            if (s == null)
                return false;

            if (s.Count != Count)
                return false;

            return ContainsAll(s);
        }

        public override int GetHashCode()
        {
            if (pset != null)
                return pset.GetHashCode();

            int h = 0;
            foreach (IPersistent o in this)
            {
                h += o.Oid;
            }
            return h;
        }

        public override void Deallocate()
        {
            if (pset != null)
                pset.Deallocate();

            base.Deallocate();
        }
    }
}

File: java/src15/org/nachodb/impl/ScalableSet.java
package org.nachodb.impl;
import  org.nachodb.*;
import java.util.*;

class ScalableSet<T extends IPersistent> extends PersistentCollection<T> implements IPersistentSet<T> { 
    Link<T>           link;
    IPersistentSet<T> set;

    static final int BTREE_THRESHOLD = 128;

    ScalableSet(StorageImpl storage, int initialSize) { 
        super(storage);
        if (initialSize <= BTREE_THRESHOLD) { 
            link = storage.<T>createLink(initialSize);
        } else { 
            set = storage.<T>createSet();
        }
    }

    ScalableSet() {}

    public boolean isEmpty() { 
        return size() != 0;
    }

    public int size() { 
        return link != null ? link.size() : set.size();
    }

    public void clear() { 
        if (link != null) { 
            link.clear();
            modify();
        } else { 
            set.clear();
        }
    }

    public boolean contains(Object o) {
        if (o instanceof IPersistent) { 
            IPersistent p = (IPersistent)o;
            return link != null ? link.contains(p) : set.contains(p);
        }
        return false;
    }
    
    public Object[] toArray() { 
        return link != null ? link.toArray() : set.toArray();
    }

    public <E> E[] toArray(E a[]) { 
        return link != null ? link.<E>toArray(a) : set.<E>toArray(a);
    }

    public Iterator<T> iterator() { 
        return link != null ? link.iterator() : set.iterator();
    }

    public boolean add(T obj) { 
        if (link != null) { 
            if (link.indexOf(obj) >= 0) { 
                return false;
            }
            if (link.size() == BTREE_THRESHOLD) { 
                set = getStorage().<T>createSet();
                for (int i = 0, n = link.size(); i < n; i++) { 
                    set.add(link.get(i));
                }
                link = null;
                modify();
                set.add(obj);
            } else { 
                modify();
                link.add(obj);
            }
            return true;
        } else { 
            return set.add(obj);
        }
    }

    public boolean remove(T o) { 
        if (link != null) {
            if (link.remove(o)) { 
                modify();
                return true;
            } 
            return false;
        } else { 
            return set.remove(o);
        }
    }
    
    public int hashCode() {
        int h = 0;
        Iterator<T> i = iterator();
        while (i.hasNext()) {
            h += i.next().getOid();
        }
        return h;
    }

    public boolean equals(Object o) {
        if (o == this) {
            return true;
        }
        if (!(o instanceof Set)) {
            return false;
        }
        Collection c = (Collection) o;
        if (c.size() != size()) {
            return false;
        }
        return containsAll(c);
    }

    public void deallocate() { 
        if (set != null) { 
            set.deallocate();
        }
        super.deallocate();
    }
}
   